src/FindIntrinsics.cpp

```diff
@@ -0,0 +1,714 @@
+#include "FindIntrinsics.h"
+#include "CodeGen_Internal.h"
+#include "ConciseCasts.h"
+#include "IRMatch.h"
+#include "IRMutator.h"
+#include "Simplify.h"
+
+namespace Halide {
+namespace Internal {
+
+using namespace Halide::ConciseCasts;
+
+namespace {
+
+bool find_intrinsics_for_type(const Type &t) {
+    // Currently, we only try to find and replace intrinsics for vector types.
+    return t.is_vector();
+}
+
+Expr lossless_narrow(const Expr &x) {
+    return lossless_cast(x.type().narrow(), x);
+}
+
+Expr strip_widening_cast(const Expr &x) {
+    Expr narrow = lossless_cast(x.type().narrow(), x);
+    if (narrow.defined()) {
+        return narrow;
+    }
+    return lossless_cast(x.type().narrow().with_code(halide_type_uint), x);
+}
+
+Expr saturating_narrow(const Expr &a) {
+    Type narrow = a.type().narrow();
+    return saturating_cast(narrow, a);
+}
+
+Expr make_shift_right(const Expr &a, int const_b) {
+    internal_assert(const_b > 0);
+    Expr b = make_const(a.type().with_code(halide_type_uint), const_b);
+    return Call::make(a.type(), Call::shift_right, {a, b}, Call::PureIntrinsic);
+}
+
+Expr make_rounding_shift_right(const Expr &a, int const_b) {
+    internal_assert(const_b > 0);
+    Expr b = make_const(a.type().with_code(halide_type_uint), const_b);
+    return Call::make(a.type(), Call::rounding_shift_right, {a, b}, Call::PureIntrinsic);
+}
+
+// Returns true iff t is an integral type where overflow is undefined
+bool no_overflow_int(Type t) {
+    return t.is_int() && t.bits() >= 32;
+}
+
+// Returns true iff t does not have a well defined overflow behavior.
+bool no_overflow(Type t) {
+    return t.is_float() || no_overflow_int(t);
+}
+
+// If there's a widening add or subtract in the first e.type().bits() / 2 - 1
+// levels down a tree of adds or subtracts, we know there's enough headroom for
+// another add without overflow.
+bool is_safe_for_add(const Expr &e, int max_depth) {
+    if (max_depth-- <= 0) {
+        return false;
+    }
+    if (const Add *add = e.as<Add>()) {
+        return is_safe_for_add(add->a, max_depth) || is_safe_for_add(add->b, max_depth);
+    } else if (const Sub *sub = e.as<Sub>()) {
+        return is_safe_for_add(sub->a, max_depth) || is_safe_for_add(sub->b, max_depth);
+    } else if (const Cast *cast = e.as<Cast>()) {
+        if (cast->type.bits() > cast->value.type().bits()) {
+            return true;
+        } else if (cast->type.bits() == cast->value.type().bits()) {
+            return is_safe_for_add(cast->value, max_depth);
+        }
+    } else if (Call::as_intrinsic(e, {Call::widening_add, Call::widening_sub})) {
+        return true;
+    }
+    return false;
+}
+
+bool is_safe_for_add(const Expr &e) {
+    return is_safe_for_add(e, e.type().bits() / 2 - 1);
+}
+
+// We want to find and remove an add of 'round' from e.
+Expr find_and_subtract(const Expr &e, const Expr &round) {
+    if (const Add *add = e.as<Add>()) {
```

Comment:
I think it would be more similar to related code if this returned Expr() on failure instead of checking same_as. I worry a branch will get added in future that happens to make a new Expr but doesn't actually successfully subtract something (e.g. someone throws in a call to simplify).